crates/runtime/src/request/table_provider.rs

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{borrow::Cow, sync::Arc};
+
+use datafusion::{
+    catalog::{Session, TableProvider},
+    common::{Constraints, Statistics},
+    error::Result as DataFusionResult,
+    physical_plan::ExecutionPlan,
+};
+use datafusion_expr::{Expr, LogicalPlan, TableProviderFilterPushDown, TableType, dml::InsertOp};
+use opentelemetry::KeyValue;
+
+use crate::request::{AsyncMarker, RequestContext};
+
+/// [`DimensionTrackedTableProvider`] enables arbitrary metric tracking (for a given set of [`RequestContext`] dimensions), for when an arbitrary [`TableProvider`] is scanned (i.e. on [`TableProvider::scan`]).
+pub struct DimensionTrackedTableProvider<F: Fn(&[KeyValue]) + Send + Sync> {
+    pub inner: Arc<dyn TableProvider>,
+    pub track: F,
+}
+
+impl<F: Fn(&[KeyValue]) + Send + Sync> std::fmt::Debug for DimensionTrackedTableProvider<F> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.inner.fmt(f)
+    }
+}
+
+#[async_trait::async_trait]
+impl<F: Fn(&[KeyValue]) + Send + Sync> TableProvider for DimensionTrackedTableProvider<F> {
```

Comment:
I'm hesitant to add yet another wrapped table provider - why can't we just add the tracking specifically to each table provider that needs it?

Reply:
Also if we did proceed with this approach, we need to make sure to implement all of the TableProvider methods, not just the required ones.

Reply:
The key issue is that we want to track a specific metric at `TableProvider::scan`, but at that time, we don't know which metric we should emit (i.e. FTS or vector search). I suppose we could emit this metric at the construction of the `TableProvider`, if that sufficiently equivalent?